### ISSUE TITLE
DPO 标签 shift

### ISSUE THREAD

**Issue by lybugproducer** [wyf3/llm_related#33]
[https://github.com/wyf3/llm_related/blob/main/train_llm_from_scratch/dataset.py](url)

第 166 行

```
        for feature in features:
            inputs_ids.append(feature[0] + feature[1])
            labels.append([0]*len(feature[0]) + feature[1])
        for feature in features:
            inputs_ids.append(feature[0] + feature[2])
            labels.append([0]*len(feature[0]) + feature[2])
```

此处没有对 labels 进行移位，会导致后面的处理错位：

[https://github.com/wyf3/llm_related/blob/main/train_llm_from_scratch/dpo_train.py](url)

第 20 行

```
    for logit, label in zip(logits, labels):
        new_logits.append(logit[label != 0].sum().unsqueeze(0))

```

第 k 个 query 的标签应该是 第 k+1 个 token：

```
        for feature in features:
            inputs_ids.append(feature[0] + feature[1])
            labels.append([0]*(len(feature[0]) - 1) + feature[1])
        for feature in features:
            inputs_ids.append(feature[0] + feature[2])
            labels.append([0]*(len(feature[0]) - 1) + feature[2])
```
